From d5ee57f01bae11098b9cb2a7db1c3fa6d3c74bee Mon Sep 17 00:00:00 2001
From: Rob Evers <revers@redhat.com>
Date: Wed, 25 Jun 2014 14:11:36 -0400
Subject: [scsi] be2iscsi: Fix exposing Host in sysfs after adapter initialization is complete

Message-id: <1403705504-27660-3-git-send-email-revers@redhat.com>
Patchwork-id: 84799
O-Subject: [RHEL7.1 e-stor PATCH 02/10] scsi: be2iscsi: Fix exposing Host in sysfs after adapter initialization is complete
Bugzilla: 1088576
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Tomas Henzl <thenzl@redhat.com>
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>

From: John Soni Jose <sony.john-n@emulex.com>

http://www.spinics.net/lists/linux-scsi/msg74149.html

 Before probe for function was completed, iSCSI Daemon had initiated login
 to target while OS was coming up. The targets which had node.startup=automatic,
 login process was initiated.Since function specific initialization was still in
 progress this lead to kernel panic.

 Fixed the issue by moving iscsi_host_add() call after adapter initialization
 is done.

Signed-off-by: John Soni Jose <sony.john-n@emulex.com>
Signed-off-by: Jayamohan Kallickal <jayamohan.kallickal@emulex.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/be2iscsi/be_main.c b/drivers/scsi/be2iscsi/be_main.c
index c1d7e0f..634ad81 100644
--- a/drivers/scsi/be2iscsi/be_main.c
+++ b/drivers/scsi/be2iscsi/be_main.c
@@ -599,15 +599,7 @@ static struct beiscsi_hba *beiscsi_hba_alloc(struct pci_dev *pcidev)
  pci_set_drvdata(pcidev, phba);
  phba->interface_handle = 0xFFFFFFFF;
 
- if (iscsi_host_add(shost, &phba->pcidev->dev))
-  goto free_devices;
-
  return phba;
-
-free_devices:
- pci_dev_put(phba->pcidev);
- iscsi_host_free(phba->shost);
- return NULL;
 }
 
 static void beiscsi_unmap_pci_function(struct beiscsi_hba *phba)
@@ -5700,6 +5692,9 @@ static int beiscsi_dev_probe(struct pci_dev *pcidev,
  }
  hwi_enable_intr(phba);
 
+ if (iscsi_host_add(phba->shost, &phba->pcidev->dev))
+  goto free_blkenbld;
+
  if (beiscsi_setup_boot_info(phba))
   /*
    * log error but continue, because we may not be using
-- 
1.7.1